File: advent-of-code/advent_of_code_2016/day_5/part_2.js
const md5 = require('md5');

var input = 'abc';
var index = 0;
var password = [];
var validHashPattern = /^00000.+/;
var hash;

while (password.length < 8 || password.includes(undefined)) {
  index += 1;
  hash = md5(input + index);

  if (validHashPattern.test(hash)) {
    if (/[^0-7]/.test(hash[5])) continue;
    let position = Number(hash[5]);
    if (password[position]) continue;
    password[position] = hash[6];
  }
}

console.log(password.join(''));

File: advent-of-code/advent_of_code_2016/day_5/part_2.rb
require 'digest'
require 'benchmark'

puts Benchmark.measure {
  input = 'reyedfim'
  index = 0
  password = Array.new(8, nil)

  loop do
    index += 1
    hash = Digest::MD5.hexdigest(input + index.to_s)

    if hash.match?(/\A00000.+/)
      position = hash[5]
      next unless position.match?(/[0-7]/)
      position = position.to_i
      next unless password[position].nil?

      password[position] = hash[6]
      break unless password.include?(nil)
    end
  end

  p password.join
}
